tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc

```diff
@@ -2797,5 +2980,563 @@ llvm::Optional<Value> convertTFConv2DCommon(
+      tosa_values_reshape_op.getResult(), tosa_indices_reshape_op.getResult());
+
+  // Finally, reshape back to the original output shape of [Indices,
+  // ParamChannels].
+  return rewriter
+      .create<tosa::ReshapeOp>(op->getLoc(), result_type,
+                               tosa_gather_op.getResult(),
+                               rewriter.getI64ArrayAttr(result_type.getShape()))
+      .getResult();
+}
+
+// Lowers OneHot operator to a sequence of TOSA ops.
+llvm::Optional<Value> convertOneHotOp(PatternRewriter& rewriter, Operation* op,
+                                      Value result_value, Value indices_value,
+                                      Value on_value, Value off_value,
+                                      int32_t depth, int32_t axis) {
+  auto result_type = result_value.getType().dyn_cast<RankedTensorType>();
+  auto indices_type = indices_value.getType().dyn_cast<RankedTensorType>();
+  auto on_value_type = on_value.getType().dyn_cast<RankedTensorType>();
+  auto off_value_type = off_value.getType().dyn_cast<RankedTensorType>();
+
+  if (!result_type) return llvm::None;
+  if (!indices_type) return llvm::None;
+  if (!on_value_type) return llvm::None;
+  if (!off_value_type) return llvm::None;
+
+  // OneHot operator creates a new tensor with shape indices.shape[:axis] +
+  // [depth] + indices.shape[axis:] For each index in 'indices', it needs to be
+  // within range of [0, depth - 1] and the [..., k, ...] = on_value (if k =
+  // index), or [..., k, ...] = off_value (if k != index)
+  //
+  // The lowering below assumes depth is always known at compile time.
+  // TBD for depth resolved in run time.
+  //
+  // OneHot can be lowered as TOSA Scatter, where off_value being mapped to
+  // 'values_in', on_value being mapped to 'input', and indices naturally mapped
+  // to 'indices'. Also the dimensions of TOSA scatter (N, W, K, C) need to be
+  // picked.
+  //
+  // N: number of elements of input indices
+  // Computed as:
+  // product(indices.shape[:])
+  //
+  // K: newly added dimension
+  // K = depth
+  //
+  // W, C: dummy dimension now
+  // W = C = 1
+  //
+  // High level description of lowering looks like:
+  // 1. off_value is reshaped/tiled into [N, K, C]
+  // 2. on_value is reshaped/tiled into [N, W, C]
+  // 3. indices is reshaped into [N, W]
+  // 4. scatter into [N, K, C]
+  // 5. reshaped into [LeftDims, RightDims, K]
+  // 6. transpose into [LeftDims, K, RightDims]
+  // 7. reshaped to result.shape
+
+  if (on_value_type.getRank() != 0 || off_value_type.getRank() != 0) {
+    op->emitOpError("OneHotOp: on_value/off_value needs to be scalar");
+    return llvm::None;
+  }
+
+  if (axis < -1 || axis > indices_type.getRank()) {
+    op->emitOpError("OneHotOp: axis out of valie range [-1, indices.rank]");
+    return llvm::None;
+  }
+
+  // axis = -1 is equivalent to axis = indices.rank
+  if (axis == -1) {
+    axis = indices_type.getRank();
+  }
+
+  int32_t N = 1, W = 1, C = 1;
```

Comment:
Prefer just `int` instead of `int32_t` unless if there is a strong reason (here and below).

Reply:
Thanks will update in next round.